samples/guestbook/main.go

```diff
@@ -107,14 +107,18 @@ func main() {
 	defer cleanup()
 
 	// Set up URL routes.
-	r := mux.NewRouter()
+	r := app.srv.Handler
+	// r := mux.NewRouter()
+
+	// these all error, saying http.Handler has no method
+	// so is the mux.NewRouter actually getting injected?
 	r.HandleFunc("/", app.index)
```

Comment:
These three lines all error with 
```
./main.go:118:3: r.HandleFunc undefined (type http.Handler has no field or method HandleFunc)
./main.go:119:3: r.HandleFunc undefined (type http.Handler has no field or method HandleFunc)
./main.go:120:3: r.HandleFunc undefined (type http.Handler has no field or method HandleFunc)
```

Can anyone help me figure out why the gorrilla/mux router is not actually getting injected? (It is in the wiregen file, see: [other comment](https://github.com/google/go-cloud/pull/1751/files#r271899065) @shantuo @jba @vangent

Reply:
Sorry, I'm not familiar with wire.

Reply:
I don't think this is a wire problem. Before your change, `r` was a `mux.Router`; after this, r is a `http.Handler`. It looks like you are expecting `wire.Bind` to inject the mux.Router into the `src.Handler` field, and maybe that works because mux.Router implements `http.Handler`? But in this code, the type is still `http.Handler`, which doesn't have a `HandleFunc` method.
HTH

Reply:
yeah, `mux.Router` implements `http.Handler`. If you look at the other comments, you can see that the `router` is the `mux.NewRouter()` and it's getting passed to the server constructor. Then on [old line 94, new line 96](https://github.com/google/go-cloud/pull/1751/files#diff-f3a898cf69ea8307808637876beeefccR96) it's getting passed to the construction of the application. There are similar lines for all of the 'setup' functions. The `app` struct is getting created with those setup functions, in every case except for the `log.Fatalf` case, [which exits](https://golang.org/pkg/log/#Fatalf). So as far as I can tell, there's no code path that doesn't get injected with the `mux.Router`, so there's no way for it to ever be the zero type of `http.Handler` there, but I am still getting this error.